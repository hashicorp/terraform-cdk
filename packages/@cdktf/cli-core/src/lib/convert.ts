/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

import * as hcl2cdk from "@cdktf/hcl2cdk";
import { CdktfConfig } from "./cdktf-config";
import {
  ConstructsMakerProviderTarget,
  TerraformProviderConstraint,
  readSchema,
} from "@cdktf/provider-generator";
import { Errors, LANGUAGES } from "@cdktf/commons";

export async function convertConfigurationFile(configuration: string) {
  const cfg = CdktfConfig.read(process.cwd()); // TODO: make this the project directory instead of cwd
  const targets = cfg.terraformProviders.map((constraint) =>
    ConstructsMakerProviderTarget.from(
      new TerraformProviderConstraint(constraint),
      LANGUAGES[0]
    )
  );
  const { providerSchema } = await readSchema(targets);
  if (!providerSchema) {
    throw Errors.Internal("Could not fetch provider schema");
  }

  const { all } = await hcl2cdk.convert(configuration, {
    providerSchema,
    language: cfg.language,
    codeContainer: "constructs.Construct",
  });

  return all;
}
