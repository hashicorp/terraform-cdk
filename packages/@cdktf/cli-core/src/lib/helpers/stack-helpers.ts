// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0

import { ensureAllSettledBeforeThrowing, Errors, logger } from "@cdktf/commons";
import minimatch from "minimatch";
import { CdktfStack } from "../cdktf-stack";
import { SynthesizedStack } from "../synth-stack";

export function getSingleStack(
  stacks: SynthesizedStack[],
  stackName?: string,
  targetAction?: string,
) {
  if (!stacks) {
    throw Errors.Internal(
      "Trying to access a stack before it has been synthesized",
    );
  }

  if (stackName) {
    const stack = stacks.find((s) => s.name === stackName);
    if (!stack) {
      throw Errors.Usage("Could not find stack: " + stackName);
    }

    return stack;
  }

  if (stacks.length === 1) {
    return stacks[0];
  }

  throw Errors.Usage(
    `Found more than one stack, please specify a target stack. Run cdktf ${
      targetAction || "<verb>"
    } <stack> with one of these stacks: ${stacks
      .map((s) => s.name)
      .join(", ")} `,
  );
}

export function getMultipleStacks(
  stacks: SynthesizedStack[],
  patterns?: string[],
  targetAction?: string,
) {
  if (!patterns || !patterns.length) {
    if (stacks.length === 1) {
      return [stacks[0]];
    }
    throw Errors.Usage(
      `Found more than one stack, please specify a target stack. Run cdktf ${
        targetAction || "<verb>"
      } <stack> with one of these stacks: ${stacks
        .map((s) => s.name)
        .join(", ")} `,
    );
  }

  return patterns.flatMap((pattern) => {
    const matchingStacks = stacks.filter((stack) =>
      minimatch(stack.name, pattern),
    );

    if (matchingStacks.length === 0) {
      throw Errors.Usage(`Could not find stack for pattern '${pattern}'`);
    }

    return matchingStacks;
  });
}

// Returns the first stack that has no unmet dependencies
// An unmet dependency is a dependency that has not been deployed yet
// If there is no unfinished stack, returns undefined
// If there is no stack ready to be worked on, it returns a promise that will resolve as soon as there is a follow-up stack available
export async function getStackWithNoUnmetDependencies(
  stackExecutors: CdktfStack[],
): Promise<CdktfStack | undefined> {
  logger.debug("Checking for stacks with unmet dependencies");
  logger.debug("stack executors:", stackExecutors);
  const pendingStacks = stackExecutors.filter((executor) => executor.isPending);
  logger.debug("pending stacks:", stackExecutors);

  if (pendingStacks.length === 0) {
    return undefined;
  }

  const currentlyReadyStack = pendingStacks.find((executor) =>
    executor.stack.dependencies.every(
      (dependency) =>
        stackExecutors.find((executor) => executor.stack.name === dependency)
          ?.currentState === "done",
    ),
  );

  if (currentlyReadyStack) {
    logger.debug("Found a stack with no unmet dependencies");
    return currentlyReadyStack;
  }

  const stackExecutionPromises = stackExecutors
    .filter((ex) => ex.currentWorkPromise)
    .map((ex) => ex.currentWorkPromise);

  logger.debug(
    `${stackExecutionPromises.length} stacks are currently busy, waiting for one to finish`,
  );

  if (!stackExecutionPromises.length) {
    return undefined;
  }

  await ensureAllSettledBeforeThrowing(
    Promise.race(stackExecutionPromises),
    stackExecutionPromises,
  );

  return await getStackWithNoUnmetDependencies(stackExecutors);
}

function findAllDependantStacks(
  stackExecutors: CdktfStack[],
  stackName: string,
): CdktfStack[] {
  return stackExecutors.filter((innerExecutor) =>
    innerExecutor.stack.dependencies.includes(stackName),
  );
}

export function findAllNestedDependantStacks(
  stackExecutors: CdktfStack[],
  stackName: string,
  knownDependantStackNames: Set<string> = new Set(),
): CdktfStack[] {
  const dependantStacks = findAllDependantStacks(stackExecutors, stackName);
  dependantStacks.forEach((stack) => {
    if (knownDependantStackNames.has(stack.stack.name)) {
      return;
    }

    knownDependantStackNames.add(stack.stack.name);
    findAllNestedDependantStacks(
      stackExecutors,
      stack.stack.name,
      knownDependantStackNames,
    );
  });

  return stackExecutors.filter((executor) =>
    knownDependantStackNames.has(executor.stack.name),
  );
}

// Returns the first stack that has no dependents that need to be destroyed first
export async function getStackWithNoUnmetDependants(
  stackExecutors: CdktfStack[],
): Promise<CdktfStack | undefined> {
  logger.debug("Checking for stacks with unmet dependants");
  logger.debug("stack executors:", stackExecutors);
  const pendingStacks = stackExecutors.filter((executor) => executor.isPending);
  logger.debug("pending stacks:", stackExecutors);

  if (pendingStacks.length === 0) {
    return undefined;
  }

  const currentlyReadyStack = pendingStacks.find((executor) => {
    const dependantStacks = findAllDependantStacks(
      stackExecutors,
      executor.stack.name,
    );
    return dependantStacks.every((stack) => stack.currentState === "done");
  });

  if (currentlyReadyStack) {
    logger.debug("Found a stack with no unmet dependants");
    return currentlyReadyStack;
  }
  const stackExecutionPromises = stackExecutors
    .filter((ex) => ex.currentWorkPromise)
    .map((ex) => ex.currentWorkPromise);

  logger.debug(
    `${stackExecutionPromises.length} stacks are currently busy, waiting for one to finish`,
  );
  if (!stackExecutionPromises.length) {
    return undefined;
  }

  await Promise.race(stackExecutionPromises);
  return await getStackWithNoUnmetDependants(stackExecutors);
}

// Throws an error if there is a dependant stack that is not included
export function checkIfAllDependantsAreIncluded(
  stacksToRun: SynthesizedStack[],
  allStacks: SynthesizedStack[],
) {
  const allDependants = new Set<string>();
  stacksToRun
    .map((stack) =>
      allStacks.filter((s) => s.dependencies.includes(stack.name)),
    )
    .flat()
    .forEach((dependant) => allDependants.add(dependant.name));

  const stackNames = stacksToRun.map((stack) => stack.name);
  const missingDependants = [...allDependants].filter(
    (dependant) => !stackNames.includes(dependant),
  );

  if (missingDependants.length > 0) {
    throw Errors.Usage(
      `The following dependant stacks are not included in the stacks to run: ${missingDependants.join(
        ", ",
      )}. Either add them or add the --ignore-missing-stack-dependencies flag.`,
    );
  }
}

/** 
  Throws an error if there is a dependency that is not included
  Cycles are detected on dependency creation at synthesis time
  Running this prevents us from being in a situation where we 
  have to wait for a stack to be deployed that is not included to be run
*/
export function checkIfAllDependenciesAreIncluded(
  stacksToRun: SynthesizedStack[],
) {
  const allDependencies = new Set<string>();
  stacksToRun
    .map((stack) => stack.dependencies)
    .flat()
    .forEach((dependency) => allDependencies.add(dependency));

  const stackNames = stacksToRun.map((stack) => stack.name);
  const missingDependencies = [...allDependencies].filter(
    (dependency) => !stackNames.includes(dependency),
  );

  if (missingDependencies.length > 0) {
    throw Errors.Usage(
      `The following dependencies are not included in the stacks to run: ${missingDependencies.join(
        ", ",
      )}. Either add them or add the --ignore-missing-stack-dependencies flag.`,
    );
  }
}
