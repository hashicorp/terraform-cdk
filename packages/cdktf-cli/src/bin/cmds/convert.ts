// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import yargs from "yargs";
import { Errors } from "@cdktf/commons";
import { BaseCommand } from "./helper/base-command";
import { requireHandlers } from "./helper/utilities";
import * as path from "path";
import * as fs from "fs-extra";

function readCdktfJson(cwd = process.cwd()): { language: string } | undefined {
  try {
    const cdktfJsonPath = path.join(cwd, "cdktf.json");
    return fs.readJsonSync(cdktfJsonPath);
  } catch (e) {
    return undefined;
  }
}

class Command extends BaseCommand {
  public readonly command = "convert";
  public readonly describe =
    "Converts a single file of HCL configuration to CDK for Terraform. Takes the file to be converted on stdin.";

  public readonly builder = (args: yargs.Argv) =>
    args
      .example(
        "cat main.tf | cdktf convert  --provider integrations/github",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content and prints it"
      )
      .example(
        "cat main.tf | cdktf convert --provider hashicorp/aws > imported.ts",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content in imported.ts"
      )
      .example(
        "cat main.tf | cdktf convert --provider 'hashicorp/aws@ ~>3.62.0' 'integrations/github@ ~>4.16.0' --language python > imported.py",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content in imported.py"
      )
      .option("language", {
        choices: ["typescript", "python", "csharp", "java", "go"],
        default: readCdktfJson()?.language || "typescript",
      })
      .option("provider", {
        describe:
          "The conversion needs to know which providers are used in addition to the ones in your cdktf.json file. We search for a cdktf.json below your current working directory.",
        type: "array",
        default: [],
      })
      .option("stack", {
        describe: "Wrap the generated code within a stack class",
        type: "boolean",
        default: false,
      })
      .showHelpOnFail(true);

  public async handleCommand(argv: any) {
    Errors.setScope("convert");
    // deferred require to keep cdktf-cli main entrypoint small (e.g. for fast shell completions)
    const api = requireHandlers();
    if (argv.withProject !== undefined) {
      if (argv.language === "typescript") {
        argv.withProject = false;
      } else {
        argv.withProject = true;
      }
    }
    await api.convert(argv);
  }
}

module.exports = new Command();
