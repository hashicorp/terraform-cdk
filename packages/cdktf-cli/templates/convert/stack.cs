using System;
using Constructs;
using HashiCorp.Cdktf;
/* {{ imports }} */

namespace MyCompany.MyApp
{
    class MyStack : TerraformStack
    {
        public MyStack(Construct scope, string id) : base(scope, id)
        {
            /* {{ code }} */
        }
    }
}