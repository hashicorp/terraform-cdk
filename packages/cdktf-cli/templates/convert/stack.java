package com.mycompany.app;

import software.constructs.Construct;
import com.hashicorp.cdktf.TerraformStack;
/* {{ imports }} */

public class MyStack extends TerraformStack
{
    public MyStack(final Construct scope, final String id) {
        super(scope, id);

        /* {{ code }} */
    }
}