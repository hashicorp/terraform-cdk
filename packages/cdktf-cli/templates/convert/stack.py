#!/usr/bin/env python
# Copyright (c) HashiCorp, Inc.
# SPDX-License-Identifier: MPL-2.0

from constructs import Construct
from cdktf import TerraformStack
# {{ imports }}


class MyStack(TerraformStack):
    def __init__(self, scope: Construct, id: str):
        super().__init__(scope, id)

        # {{ code }}
