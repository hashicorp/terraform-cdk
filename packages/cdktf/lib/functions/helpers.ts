// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { Tokenization, Token } from "../tokens/token";
import { call } from "../tfExpression";
import { IResolvable } from "../tokens/resolvable";
import { TokenString, extractTokenDouble } from "../tokens/private/encoding";

type TFValue<T> = { variadic?: boolean; value: T };
type TFValueValidator<T> = (value: T) => TFValue<T>;

type ExecutableTfFunction = (...args: any[]) => IResolvable;

/**
 * Determines if given str has unescaped double quotes
 * @param str String to test
 * @returns Boolean
 */
function hasUnescapedDoubleQuotes(str: string) {
  return /(^|[^\\])([\\]{2})*"/.test(str);
}

// Validators
// eslint-disable-next-line jsdoc/require-jsdoc
export function anyValue<T>(value: T): TFValue<T> {
  return { value };
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function mapValue<T>(value: T): TFValue<T> {
  return { value };
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function stringValue<T extends string>(value: T): TFValue<T> {
  if (typeof value !== "string" && !Tokenization.isResolvable(value)) {
    throw new Error(`'${value}' is not a valid string nor a token`);
  }

  if (typeof value === "string" && hasUnescapedDoubleQuotes(value)) {
    throw new Error(
      `'${value}' can not be used as value directly since it has unescaped double quotes in it. To safely use the value please use Fn.rawString on your string.`
    );
  }

  return { value };
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function numericValue<T>(value: T): TFValue<T> {
  if (typeof value !== "number" && !Tokenization.isResolvable(value)) {
    throw new Error(`${value} is not a valid number nor a token`);
  }
  return { value };
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function listOf<T>(type: TFValueValidator<T>): TFValueValidator<T[]> {
  return (value: any) => {
    if (Tokenization.isResolvable(value)) {
      return { value };
    }

    if (!Array.isArray(value)) {
      //   throw new Error(`${value} is not a valid list`);
      return { value };
    }

    return {
      value: value
        .filter((item) => item !== undefined && item !== null)
        .map((item, i) => {
          if (Tokenization.isResolvable(item)) {
            return item;
          }

          if (TokenString.forListToken(item).test()) {
            return item;
          }

          if (extractTokenDouble(item, true) !== undefined) {
            return item;
          }

          if (TokenString.forMapToken(item).test()) {
            return item;
          }

          if (typeof item === "string") {
            const tokenList = Tokenization.reverseString(item);
            const numberOfTokens =
              tokenList.tokens.length + tokenList.intrinsic.length;
            if (numberOfTokens === 1 && tokenList.literals.length === 0) {
              return item;
            }
          }

          try {
            type(item);
            return typeof item === "string" ? `"${item}"` : item;
          } catch (error) {
            throw new Error(
              `Element in list ${value} at position ${i} is not of the right type: ${error}`
            );
          }
        }),
    };
  };
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function variadic<T>(type: TFValueValidator<T>): TFValueValidator<T[]> {
  return (value: any) => {
    // we use the list validator and set variadic to true in order to have it expanded in the args passed to the TF function
    return { value: listOf(type)(value).value, variadic: true };
  };
}

// Tokenization
// eslint-disable-next-line jsdoc/require-jsdoc
export function asString(value: IResolvable) {
  return Token.asString(value);
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function asNumber(value: IResolvable) {
  return Token.asNumber(value);
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function asList(value: IResolvable) {
  return Token.asList(value);
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function asStringMap(value: IResolvable) {
  return Token.asStringMap(value);
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function asBoolean(value: IResolvable) {
  return value; // Booleans can not be represented as a token
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function asAny(value: IResolvable) {
  // Ordinarily casting to any can cause issues, but
  // in this case it makes using functions a bit easier in TS
  // and doesn't really harm other languages.
  // Jsii has issues when returning the value directly,
  // so wrap as a string.
  return asString(value) as any;
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function terraformFunction(
  name: string,
  argValidators: TFValueValidator<any>[]
): ExecutableTfFunction {
  return function (...args: any[]) {
    if (args.length !== argValidators.length) {
      throw new Error(
        `${name} takes ${argValidators.length} arguments, but ${args.length} were provided`
      );
    }
    return call(
      name,
      // We use flatMap now, since listOf() wraps everything in an extra array but variadic() does not
      // This enables us to have variadic pass multiple args into the call() from a single array
      args.reduce((carry, arg, i) => {
        try {
          const val = argValidators[i](arg);
          if (val.variadic) return [...carry, ...val.value];
          else return [...carry, val.value];
        } catch (error) {
          throw new Error(
            `Argument ${i} of ${name} failed the validation: ${error}`
          );
        }
      }, [])
    );
  };
}
