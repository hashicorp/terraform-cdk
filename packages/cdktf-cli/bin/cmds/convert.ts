import yargs from "yargs";
import { Errors } from "../../lib/errors";
import { requireHandlers } from "./helper/utilities";

class Command implements yargs.CommandModule {
  public readonly command = "convert [OPTIONS]";
  public readonly describe =
    "Converts a single file of HCL configuration to CDK for Terraform. Takes the file to be converted on stdin.";

  public readonly builder = (args: yargs.Argv) =>
    args
      .example(
        "cat main.tf | cdktf convert  --provider integrations/github",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content and prints it"
      )
      .example(
        "cat main.tf | cdktf convert --provider hashicorp/aws > imported.ts",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content in imported.ts"
      )
      .example(
        "cat main.tf | cdktf convert --provider 'hashicorp/aws@ ~>3.62.0' 'integrations/github@ ~>4.16.0' --language python > imported.py",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content in imported.ts"
      )
      .option("language", {
        choices: ["typescript", "python", "csharp", "java"],
        default: "typescript",
      })
      .option("provider", {
        describe:
          "The conversion needs to know which providers are used in addition to the ones in your cdktf.json file. We search for a cdktf.json below your current working directory.",
        type: "array",
        default: [],
      })
      .showHelpOnFail(true);

  public async handler(argv: any) {
    Errors.setScope("convert");
    // deferred require to keep cdktf-cli main entrypoint small (e.g. for fast shell completions)
    const api = requireHandlers();

    await api.convert(argv);
  }
}

module.exports = new Command();
