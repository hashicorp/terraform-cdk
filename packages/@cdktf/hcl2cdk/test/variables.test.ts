/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

import { testCase, Synth, binding, Snapshot } from "./helpers/convert";

describe("variables", () => {
  testCase.test(
    "basic",
    `
    provider "local" {}
      variable "image_id" {
        type = string
      }
      data "local_file" "foo" {
        filename = "./\${var.image_id}.img"
      }
      `,
    [binding.local],
    Snapshot.yes,
    Synth.yes
  );

  testCase.test(
    "complex type and default",
    `
    provider "local" {}
    variable "docker_ports" {
        type = list(object({
            internal = number
            external = number
            protocol = string
        }))
        default = [
            {
            internal = 8300
            external = 8300
            protocol = "tcp"
            }
        ]
        }

        data "local_file" "foo" {
            filename = "./\${var.docker_ports[0].protocol}.img"
          }`,
    [binding.local],
    Snapshot.yes,
    Synth.yes
  );

  testCase.test(
    "sensitive and required",
    `
    provider "local" {}
    variable "user_information" {
        type = object({
          name    = string
          address = string
        })
        sensitive = true
        nullable = false
      }

      data "local_file" "foo" {
        filename = "./\${var.user_information.name}.img"
      }
      `,
    [binding.local],
    Snapshot.yes,
    Synth.yes
  );

  testCase.test(
    "validations",
    `
    provider "local" {}
    variable "image_id" {
        type        = string
        description = "The id of the machine image (AMI) to use for the server."
      
        validation {
          condition     = length(var.image_id) > 4 && substr(var.image_id, 0, 4) == "ami-"
          error_message = "The image_id value must be a valid AMI id, starting with ami-."
        }
      }

      data "local_file" "foo" {
        filename = "./\${var.image_id}.img"
      }
      `,
    [binding.local],
    Snapshot.yes,
    Synth.yes
  );
});
