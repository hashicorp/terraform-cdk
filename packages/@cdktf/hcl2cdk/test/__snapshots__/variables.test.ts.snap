// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`variables basic snapshot typescript 1`] = `
"import { Construct } from "constructs";
import { VariableType, TerraformVariable, TerraformStack } from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { DataLocalFile } from "./.gen/providers/local/data-local-file";
import { LocalProvider } from "./.gen/providers/local/provider";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    new LocalProvider(this, "local", {});
    const imageId = new TerraformVariable(this, "image_id", {
      type: VariableType.STRING,
    });
    new DataLocalFile(this, "foo", {
      filename: "./\${" + imageId.value + "}.img",
    });
  }
}
"
`;

exports[`variables complex type and default snapshot typescript 1`] = `
"import { Construct } from "constructs";
import {
  VariableType,
  TerraformVariable,
  Fn,
  Token,
  TerraformStack,
} from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { DataLocalFile } from "./.gen/providers/local/data-local-file";
import { LocalProvider } from "./.gen/providers/local/provider";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    new LocalProvider(this, "local", {});
    const dockerPorts = new TerraformVariable(this, "docker_ports", {
      default: [
        {
          external: 8300,
          internal: 8300,
          protocol: "tcp",
        },
      ],
      type: VariableType.list(
        VariableType.object({
          external: VariableType.NUMBER,
          internal: VariableType.NUMBER,
          protocol: VariableType.STRING,
        })
      ),
    });
    new DataLocalFile(this, "foo", {
      filename:
        "./" +
        Token.asString(Fn.lookupNested(dockerPorts.value, ["0", "protocol"])) +
        ".img",
    });
  }
}
"
`;

exports[`variables sensitive and required snapshot typescript 1`] = `
"import { Construct } from "constructs";
import {
  VariableType,
  TerraformVariable,
  Fn,
  Token,
  TerraformStack,
} from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { DataLocalFile } from "./.gen/providers/local/data-local-file";
import { LocalProvider } from "./.gen/providers/local/provider";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    new LocalProvider(this, "local", {});
    const userInformation = new TerraformVariable(this, "user_information", {
      sensitive: true,
      nullable: false,
      type: VariableType.object({
        address: VariableType.STRING,
        name: VariableType.STRING,
      }),
    });
    new DataLocalFile(this, "foo", {
      filename:
        "./" +
        Token.asString(Fn.lookupNested(userInformation.value, ["name"])) +
        ".img",
    });
  }
}
"
`;

exports[`variables validations snapshot typescript 1`] = `
"import { Construct } from "constructs";
import { VariableType, TerraformVariable, TerraformStack } from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { DataLocalFile } from "./.gen/providers/local/data-local-file";
import { LocalProvider } from "./.gen/providers/local/provider";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    new LocalProvider(this, "local", {});
    const imageId = new TerraformVariable(this, "image_id", {
      description: "The id of the machine image (AMI) to use for the server.",
      validation: [
        {
          errorMessage:
            "The image_id value must be a valid AMI id, starting with ami-.",
          condition:
            '\${length(var.image_id) > 4 && substr(var.image_id, 0, 4) == "ami-"}',
        },
      ],
      type: VariableType.STRING,
    });
    new DataLocalFile(this, "foo", {
      filename: "./\${" + imageId.value + "}.img",
    });
  }
}
"
`;
